samples/UpgradeStepSample/UpgradeStepSample/SampleExtensionServiceProvider.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Microsoft.DotNet.UpgradeAssistant;
+using Microsoft.DotNet.UpgradeAssistant.Extensions;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace UpgradeStepSample
+{
+    /// <summary>
+    /// Extension authors can implement the IExtensionServiceProvider interface to
+    /// register services with Upgrade Assistant's dependency injection container.
+    /// This could include registering additional upgrade steps. It might also include
+    /// registering services needed by the steps registered or other migrations steps.
+    /// For example, registering Roslyn analyzer/code fix providers, IConfigUpdaters,
+    /// or IPackageReferenceAnalyzers will cause upgrade steps that use those types to
+    /// pick the newly registered services up automatically and use them.
+    /// </summary>
+    public class SampleExtensionServiceProvider : IExtensionServiceProvider
+    {
+        /// <summary>
+        /// Registers services needed for the AuthorsProperty sample extension in
+        /// Upgrade Assistant's dependency injection container.
+        /// </summary>
+        /// <param name="serviceConfiguration">A configuration object containing both the
+        /// service collection to register services in and the extension's configuration file.</param>
+        /// <returns>The service collection updated with services Upgrade Assistant should use.</returns>
+        public IServiceCollection AddServices(ExtensionServiceConfiguration serviceConfiguration)
+        {
+            if (serviceConfiguration is null)
+            {
+                throw new ArgumentNullException(nameof(serviceConfiguration));
+            }
+
+            return serviceConfiguration.ServiceCollection.AddScoped<UpgradeStep, AuthorsPropertyUpgradeStep>();
```

Comment:
Can we add a method of `AddStep<>`?  Just to ensure steps are added with the right scope? As far as scope, do we need to to be scoped or would transient be fine?

Reply:
Good idea. The scope question is interesting. Transient should be fine, I think. I think the reason behind them originally having a scoped lifetime was that the upgrade steps may store state (between init and apply) and I wanted to make sure that different services resolving a step within that scenario would get the same step. So there's theoretically some use in the scope lifetime. In practice, though, I think we only resolve them once and then just pass them around, so transient is probably fine.